alvr/client_openxr/src/extra_extensions/multimodal_input.rs

```diff
@@ -14,47 +14,104 @@ pub const META_DETACHED_CONTROLLERS_EXTENSION_NAME: &str = "XR_META_detached_con
+    session: xr::Session<xr::AnyGraphics>,
+    pause_simultaneous_hands_and_controllers_tracking_meta:
+        PauseSimultaneousHandsAndControllersTrackingMETA,
+}
 
-pub fn resume_simultaneous_hands_and_controllers_tracking<G>(
-    session: &xr::Session<G>,
-) -> xr::Result<()> {
-    let resume_simultaneous_hands_and_controllers_tracking_meta = unsafe {
-        let mut resume_simultaneous_hands_and_controllers_tracking_meta = None;
-        let _ = (session.instance().fp().get_instance_proc_addr)(
-            session.instance().as_raw(),
-            c"xrResumeSimultaneousHandsAndControllersTrackingMETA".as_ptr(),
-            &mut resume_simultaneous_hands_and_controllers_tracking_meta,
-        );
-
-        resume_simultaneous_hands_and_controllers_tracking_meta.map(|pfn| {
-            mem::transmute::<VoidFunction, ResumeSimultaneousHandsAndControllersTrackingMETA>(pfn)
+impl MultimodalHandsHandle {
+    pub fn new<G>(session: xr::Session<G>) -> xr::Result<Self> {
+        let session = session.into_any_graphics();
+
+        let resume_simultaneous_hands_and_controllers_tracking_meta = unsafe {
+            let mut resume_simultaneous_hands_and_controllers_tracking_meta = None;
+            let _ = (session.instance().fp().get_instance_proc_addr)(
+                session.instance().as_raw(),
+                c"xrResumeSimultaneousHandsAndControllersTrackingMETA".as_ptr(),
+                &mut resume_simultaneous_hands_and_controllers_tracking_meta,
+            );
+
+            resume_simultaneous_hands_and_controllers_tracking_meta.map(|pfn| {
+                mem::transmute::<VoidFunction, ResumeSimultaneousHandsAndControllersTrackingMETA>(
+                    pfn,
+                )
+            })
+        }
+        .ok_or(sys::Result::ERROR_EXTENSION_NOT_PRESENT)?;
+
+        let pause_simultaneous_hands_and_controllers_tracking_meta = unsafe {
+            let mut pause_simultaneous_hands_and_controllers_tracking_meta = None;
+            let _ = (session.instance().fp().get_instance_proc_addr)(
+                session.instance().as_raw(),
+                c"xrPauseSimultaneousHandsAndControllersTrackingMETA".as_ptr(),
+                &mut pause_simultaneous_hands_and_controllers_tracking_meta,
+            );
+
+            pause_simultaneous_hands_and_controllers_tracking_meta.map(|pfn| {
+                mem::transmute::<VoidFunction, PauseSimultaneousHandsAndControllersTrackingMETA>(
+                    pfn,
+                )
+            })
+        }
+        .ok_or(sys::Result::ERROR_EXTENSION_NOT_PRESENT)?;
+
+        let resume_info = SimultaneousHandsAndControllersTrackingResumeInfoMETA {
+            ty: *TYPE_SIMULTANEOUS_HANDS_AND_CONTROLLERS_TRACKING_RESUME_INFO_META,
+            next: ptr::null(),
+        };
+        unsafe {
+            super::xr_res(resume_simultaneous_hands_and_controllers_tracking_meta(
+                session.as_raw(),
+                &resume_info,
```

Comment:
If possible i'd avoid any global state. But i agree that since the api is made of global calls, once creating two instances everything falls apart. I'm a little afraid that using a global state would still cause concurrency issues

Reply:
is it really global tho? it's specific to the session because it acts upon a specific session
if you make it so you have to pass the session into the pause/start function then that's even more emphasized
or do you have to guarantee that the function is only used with the same session the pointer was grabbed from?

Reply:
The `action_set` becomes bound to the session.
> Action sets are application-defined collections of actions. They are attached to a given [XrSession](https://registry.khronos.org/OpenXR/specs/1.1/html/xrspec.html#XrSession) with a [xrAttachSessionActionSets](https://registry.khronos.org/OpenXR/specs/1.1/html/xrspec.html#xrAttachSessionActionSets) call. They are enabled or disabled by the application via [xrSyncActions](https://registry.khronos.org/OpenXR/specs/1.1/html/xrspec.html#xrSyncActions) depending on the current application context.

Once a action set is bound to a session, if the session gets destroy, the action set needs to be recreated. But what we can do is creating multiple action sets, local to Lobby and StreamContext. This is a possible direction.

Reply:
We don't even have to modify any action sets tho? Face + body + eye tracking don't use that iirc and we always register the multimodal action set even if we don't actually unpause it, so that shouldn't be a problem

Reply:
Yeah, so if we want to implement "handles" that hold this kind of functionality we have to handle the edge cases like multimodal. So the idea would be to keep InteractionContext, but add a function to create a `InteractionSourcesHandle` that when dropped removes the requested input functionality. But with this there are still issues. IF we request multimodal for lobby, then when streaming without body tracking, multimodal will still be enabled for stream. This is because multimodal is a feature affecting the behavior of other inputs (of input/grip/pose and hand tracking active states). For this reason, in the end I believe having plain stateful `select_sources()` call is better and re responsibility gets elevated to lobby.rs, stream.rs and lib.rs

Reply:
ok so TLDR, input functionality is not always "additive", so we cannot implement independent handles to hold the extra functionality.